### DISCUSSION TITLE
Use JSON5 For package.json

### DISCUSSION THREAD

**Discussion by machineghost** [npm/rfcs#230] [RRFC]
JSON5 ("JSON for Humans"), includes features such as comments, which have long been desired in `package.json`.

If `package.json` adopted JSON5 it would add not only comments, but also a number of minor conveniences (eg. optional trailing commas, optional key quoting, single-quoted strings, etc. ... all just like in JS!)  And switching to the new standard would be incredibly simple.

See https://json5.org/ for more info.

**Comment by ljharb** [comment]
`.json` means JSON. JSON5 would necessitate a different extension.

Also, supporting JSON5 in `package.json` would break every app that passes `package.json`'s contents through `JSON.parse`, indirectly or directly.

**Comment by RemyRylan** [comment]
I think that ship sailed a long time ago. There are thousands of packages in the npm registry that expect to `require` or use `JSON.parse` for package.json. Changing to a new format would break every single one of them.

I know JSON5/JSON-C (JSON w/ comments) is gaining traction, with more tools supporting it. Still, those tools usually support it from their inception, so the supporting ecosystem is built around the requirement to parse the `.json` configs differently. TypeScript, as an example, ships with a config file parser, so you don't even have to use `json5` if you already have `typescript` installed.

But for `package.json` -- the negative impact would be too high. Best to leave as-is.

**Reply by matthew-dean** [reply]
> There are thousands of packages in the npm registry that expect to require or use JSON.parse for package.json. Changing to a new format would break every single one of them.

This just isn't true. There's no reason why the Node version of JSON.parse or require() couldn't ignore comments.

**Reply by ljharb** [reply]
@matthew-dean the JS spec dictates what `JSON.parse` does, so it would be impossible to make any changes to it like that. `require` certainly could, but most tools ime aren't using require to load package.json.

**Reply by evert** [reply]
Without having to change JSON.parse, this would be a microscopic backwards compatibility break compared to something like ESM support. Regardless, there's a standard way to introduce these changes: New major versions.

Just support .json5 at the root. Yes some packages will need to be updated. They have before, and then can again.

**Reply by ljharb** [reply]
Adding ESM to a package is almost always a major change, true, but it wasn't one for node, so I'm not sure what you mean. Zero packages ever *had* to be updated as a result of ESM support - CJS packages continue to work fine forever.

**Reply by evert** [reply]
Yes and old packages with package.json will as well

**Reply by ljharb** [reply]
… and just like packages changing to ESM is a massive breaking change, because it restricts who can consume it, packages changing package.json format would in just the same way.

Without any native JSON5 support in node, it would be impractical for any tools to update at all.

**Reply by evert** [reply]
It's such a strange way to have this discussion, but I guess I'll bite.

> Without any native JSON5 support in node, it would be impractical for any tools to update at all.

If this is a problem, it sounds like you already know a solution? That said, I wouldn't think that using a dependency for parsing JSON5 is impractical.

Anyway, everything you're mentioning are of course legitimate concerns, and would have to be taken into account. Ultimately it's up to the maintainers of this project to weigh the drawbacks with the benefits. It's also good that you're workshopping all the potential issues. Nothing you mention to me sound like deal breakers, but it's not for me to decide.

**Reply by ljharb** [reply]
The first step would be landing a native stringify/parse pair for JSON5 in node. A stronger path to that than just trying to land it there would be to try to standardize it in WinterCG; stronger than that, in browsers; and even stronger still, in the JS language.

Once that functionality was shipping in all actively supported node versions, then tooling can begin to support the format. At the same time, it would start to become reasonable for npm to explore shipping the requested functionality in some form.

In other words, it's a many year journey, involving many stakeholders and lots of compelling evidence and persuasion - but unfortunately so far nobody's been willing to put in more effort than typing up github comments, so the journey still hasn't yet begun.

**Reply by evert** [reply]
All change has to start somewhere though =). I think we can be forgiven for sharing a desire for this in the appropriate channels. I don't understand all the negativity in every comment, but maybe that's a lost in translation thing.

**Reply by wesleytodd** [reply]
Yeah I think it is just a mis-understanding of the tone. @ljharb is deeply involved in the ecosystem, specification work, and some very popular tools. I would recommend you take his responses here more as "recognization of how hard this reality can be" than "negativity".

I think a reasonable middle ground might be for someone to try and push for Node.js to vendor in a user land JSON5 parser as experimental. And then from that land experimental support for Node.js to use that file for things like `exports`. If someone wants to start that work then I think it would really help prove out this for npm to then consider it.

**Reply by evert** [reply]
Thanks, that. Sounds like a really solid first step!

**Reply by evert** [reply]
Looks like a lot of discussion has already happened here, and it's unfortunately closed:

https://github.com/nodejs/node/issues/40714

Interesting point is that pnpm does support json5.

**Comment by evert** [comment]
At the very least it would be possible to write `.json` files in when building npm packages and when installing in `node_modules`, but (optionally) support JSON5 at the top-level. Yes this will break some tools but it's an opt-in change and tooling will catch up.
